src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+        /// <param name="wrappedExpression">The expression being wrapped in a span.</param>
+        /// <param name="used">true if the result of the expression is used; false if it's required only for its side effects.</param>
+        /// <param name="inPlaceTarget">A non-null expression if the construction is initializing an existing local in-place; otherwise, null.</param>
+        /// <param name="avoidInPlace">An output Boolean indicating whether a caller trying to perform in-place initialization should instead prefer to assign the local to a new value.</param>
+        /// <param name="start">The expression for the offset into the array being wrapped in a span.</param>
+        /// <param name="length">The expression for the length of the subarray being wrapped in a span.</param>
+        /// <returns>
+        /// true if this method successfully emit a ReadOnlySpan as a wrapper for a blob; otherwise, false.  If false, nothing will have been emitted.
+        /// And if false and <paramref name="avoidInPlace"/> is true (in which case <paramref name="inPlaceTarget"/> must have been non-null), the caller
+        /// may try again but with a null <paramref name="inPlaceTarget"/>.
+        /// </returns>
+        private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExpression wrappedExpression, bool used, BoundExpression inPlaceTarget, out bool avoidInPlace, BoundExpression? start = null, BoundExpression? length = null)
         {
-            Debug.Assert(start is null == length is null);
+            Debug.Assert(start is null == length is null, "start and length must either both be null or both be non-null");
+            Debug.Assert(inPlaceTarget is null || TargetIsNotOnHeap(inPlaceTarget), "in-place construction target should not be on heap");
 
             ImmutableArray<byte> data = default;
             int elementCount = -1;
+            avoidInPlace = false;
+            SpecialType specialElementType = SpecialType.None;
 
             if (!_module.SupportsPrivateImplClass)
             {
                 return false;
             }
 
-            var ctor = (MethodSymbol?)Binder.GetWellKnownTypeMember(this._module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
-
-            if (ctor == null)
+            var rosPointerCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosPointerCtor is null)
             {
                 return false;
             }
-
-            Debug.Assert(!ctor.HasUnsupportedMetadata);
-
+            Debug.Assert(!rosPointerCtor.HasUnsupportedMetadata);
+
+            // The purpose of this optimization is to replace a BoundArrayCreation with better code generation.
```

Comment:
It would be good to mention the UTF8 scenario which you mentioned offline. That motivates why this method should handle start/length scenarios, albeit only when element size is 1.